cy43439.go

```diff
@@ -267,46 +274,84 @@ func (d *Device) handleAsyncEvent(payload []byte) error {
 	return d.processAsyncEvent(as)
 }
 
+func (d *Device) allocRecvPktQ() {
+
+	// Allocate queue of recv pkts, each pkt large enough to hold a
+	// full-sized Ethernet pkt.  Pkts on the queue are not-active and are
+	// owned by the driver.  Pkts off the queue are active and owned by the
+	// stack.
+
+	for len(d.recvPktQ) < cap(d.recvPktQ) {
+		pkt := make([]byte, 0, pktMaxSize)
+		d.recvPktQ <- pkt
+	}
+}
+
+// ref: void cyw43_cb_process_ethernet(void *cb_data, int itf, size_t len, const uint8_t *buf)
+func (d *Device) processEthernet(payload []byte) error {
+
+	println("processEthernet")
+	if d.recvEth == nil {
+		// drop packet
+		return nil
+	}
+
+	// Pop a pkt off of the queue and copy the payload into the pkt,
+	// freeing up the payload backing for the next receive.  Send the pkt
+	// up the stack.  Later, the stack will return the pkt and pushed back
+	// onto the queue.
+
+	if len(d.recvPktQ) == 0 {
+		// no pkt available, drop
+		println("DROPPING PKT")
+		return nil
+	}
+
+	pkt := <-d.recvPktQ
+	n := copy(pkt[:cap(pkt)], payload)
+	if err := d.recvEth(pkt[:n]); err != nil {
```

Comment:
We can do it that way: RecvEth cannot keep refs to the buf after it returns.  But in the stack you're still going to need to do a copy from that buf into socket bufs, which the app will read from in another context.  The copy is unavoidable, either way, because we don't want to force hw to drop a pkt because the last buf wasn't serviced in time.  So there has to be some buffering, somewhere.

The model cyw43 is using, and what I was copying, is kind of the Linux network driver model where the driver maintains a Q of bufs to receive pkts into, asynchronous from the later processing of the bufs later up the stack.  If the stack (apps) aren't servicing the bufs and returning those quick enough, the Q empties and hw drops pkts.  The sw copy cyw43 is doing could be eliminated if we had DMA working where hw would receive into the next available buf on a pre-allocated DMA ring.

You can get real fancy and avoid all sw copies using RDMA, but that brings all kinds of other problems.  But it would be the most performant (sp?) model.  Total pain in the ass, so we don't want to go there.

So how about I revert that driver buf Q and just call RecvEth with d.buf data and assume it'll released when RecvEth returns?  That way all the buffer management is pushed up into the stack.  You'll have to decide there how to pre-allocate space for socket buffers in a way that doesn't put pressure on GC.

Thoughts?

Reply:
> stack you're still going to need to do a copy [...] So there has to be some buffering, somewhere.

Agreed. I think its wise we proceed like this for now and avoid adding complexity on the drivers side which will be the hardest to debug if something goes wrong. We can rethink this API decision to eliminate the middleman copy in the future once we have a stable thing going on; at least I think this is wisest for now- simplicity+correctness on the drivers side is priority

> So how about I revert that driver buf Q and just call RecvEth with d.buf data and assume it'll released when RecvEth returns?

Yes! Lets go :smile:

Reply:
done